client/src/quic_client.rs

```diff
@@ -0,0 +1,208 @@
+    async fn _send_buffer(&self, data: &[u8]) -> Result<Arc<NewConnection>, WriteError> {
+        let connection = {
+            let mut conn_guard = self.connection.lock().await;
+
+            let maybe_conn = (*conn_guard).clone();
+            match maybe_conn {
+                Some(conn) => conn.clone(),
+                None => {
+                    let connecting = self.endpoint.connect(self.addr, "connect").unwrap();
+                    let connection = Arc::new(connecting.await?);
+                    *conn_guard = Some(connection.clone());
+                    connection
+                }
+            }
+        };
+        match Self::_send_buffer_using_conn(data, &connection).await {
+            Ok(()) => Ok(connection),
+            _ => {
+                let connection = {
+                    let connecting = self.endpoint.connect(self.addr, "connect").unwrap();
+                    let connection = Arc::new(connecting.await?);
+                    let mut conn_guard = self.connection.lock().await;
+                    *conn_guard = Some(connection.clone());
+                    connection
+                };
+                Self::_send_buffer_using_conn(data, &connection).await?;
+                Ok(connection)
+            }
+        }
+    }
+
+    pub async fn send_buffer(&self, data: &[u8]) -> Result<(), ClientErrorKind> {
+        self._send_buffer(data).await?;
+        Ok(())
+    }
+
+    pub async fn send_batch(&self, buffers: &[Vec<u8>]) -> Result<(), ClientErrorKind> {
+        // Start off by "testing" the connection by sending the first transaction
+        // This will also connect to the server if not already connected
+        // and reconnect and retry if the first send attempt failed
+        // (for example due to a timed out connection), returning an error
+        // or the connection that was used to successfully send the transaction.
+        // We will use the returned connection to send the rest of the transactions in the batch
+        // to avoid touching the mutex in self, and not bother reconnecting if we fail along the way
+        // since testing even in the ideal GCE environment has found no cases
+        // where reconnecting and retrying in the middle of a batch send
+        // (i.e. we encounter a connection error in the middle of a batch send, which presumably cannot
+        // be due to a timed out connection) has succeeded
+        if buffers.is_empty() {
+            return Ok(());
+        }
+        let connection = self._send_buffer(&buffers[0][..]).await?;
+
+        let futures = buffers[1..buffers.len()]
+            .iter()
+            .chunks(MAX_CONCURRENT_STREAMS)
+            .into_iter()
+            .map(|buffs| {
+                join_all(
+                    buffs
+                        .into_iter()
+                        .map(|buf| Self::_send_buffer_using_conn(&buf[..], &connection)),
+                )
+            })
+            .collect::<Vec<_>>();
```

Comment:
do we need the collect?  usually clippy complains when they're unnecessary, but seems like we can just loop over the iterator directly below